docs/proposals/container-init.md

```diff
@@ -0,0 +1,333 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Pod initialization
+
+@smarterclayton
+
+March 2016
+
+## Proposal and Motivation
+
+Within a pod there is a need to initialize local data or adapt to the current
+cluster environment that is not easily achieved in the current container model.
+Containers start in parallel after volumes are mounted, leaving no opportunity
+for coordination between containers without specialization of the image. If
+two containers need to share common initialization data, both images must
+be altered to cooperate using filesystem or network semantics, which introduces
+coupling between images. Likewise, if an image requires configuration in order
+to start and that configuration is environment dependent, the image must be
+altered to add the necessary templating or retrieval.
+
+This proposal introduces the concept of an **init container**, one or more
+containers started in sequence before the pod's normal containers are started.
+These init containers may share volumes, perform network operations, and perform
+computation prior to the start of the remaining containers. They may also, by
+virtue of their sequencing, block or delay the startup of application containers
+until some precondition is met. In this document we refer to the existing pod
+containers as **app containers**.
+
+## Design Points
+
+* Init containers should be able to:
+  * Perform initialization of shared volumes
+  * Delay the startup of application containers until preconditions are met
+  * Download binaries that will be used in app containers as execution targets
```

Comment:
Yes, it should be nested under that one

On Tue, Apr 12, 2016 at 5:27 PM, Vish Kannan <notifications@github.com>
wrote:

> In docs/proposals/container-init.md
> <https://github.com/kubernetes/kubernetes/pull/23666#discussion_r59456899>
> :
>
> > +altered to add the necessary templating or retrieval.
> > +
> > +This proposal introduces the concept of an **init container**, one or more
> > +containers started in sequence before the pod's normal containers are started.
> > +These init containers may share volumes, perform network operations, and perform
> > +computation prior to the start of the remaining containers. They may also, by
> > +virtue of their sequencing, block or delay the startup of application containers
> > +until some precondition is met. In this document we refer to the existing pod
> > +containers as **app containers**.
> > +
> > +## Design Points
> > +
> > +* Init containers should be able to:
> > +  * Perform initialization of shared volumes
> > +  * Delay the startup of application containers until preconditions are met
> > +  * Download binaries that will be used in app containers as execution targets
>
> Doesn't Perform initialization of shared volumes address this use case?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23666/files/7efdc3d36197f1ca452339d49abbc87163fb0e75#r59456899>
>